sql/catalyst/src/main/scala/org/apache/spark/sql/types/StructType.scala

```diff
@@ -104,6 +104,34 @@ case class StructType(fields: Array[StructField]) extends DataType with Seq[Stru
   private lazy val nameToIndex: Map[String, Int] = fieldNames.zipWithIndex.toMap
 
   /**
+   * Support construction of StructType with addition rather than providing a list.
+   */
+  def add(field: StructField): StructType = {
+    StructType(fields :+ field)
+  }
+
+  /**
+   * Support construction of StructType with addition rather than providing a list.
+   */
+  def add(name: String,
+    dataType: DataType,
+    nullable: Boolean = true,
+    metadata: Metadata = Metadata.empty): StructType = {
+    StructType(fields :+ new StructField(name, dataType, nullable, metadata))
+  }
+
+  /**
+   * Support construction of StructType with addition rather than providing a list.
+   */
+  def add(name: String,
+    dataType: String,
+    nullable: Boolean = true,
+    metadata: Metadata = Metadata.empty): StructType = {
```

Comment:
in order to support java, we cannot have default argument values. We can just create multiple versions of this though.